src/py21cmfast/wrapper/cfuncs.py

```diff
@@ -16,11 +17,82 @@
 
 logger = logging.getLogger(__name__)
 
+# Ideally, backend functions that we access here should do all the broadcasting/initialisation themselves
+# These decorators are for lower functions which are called directly in one or two lines, like delta_crit
+
+# TODO: a lot of these assume input as numpy arrays via use of .shape, explicitly require this
+
+
+def broadcast_params(func: Callable) -> Callable:
+    """Decorator to broadcast the parameters to the C library before calling the function."""
+
+    def wrapper(*args, **kwargs):
+        inputs = kwargs.get("inputs")
+        if inputs.astro_params:
+            lib.Broadcast_struct_global_all(
+                inputs.user_params.cstruct,
+                inputs.cosmo_params.cstruct,
+                inputs.astro_params.cstruct,
+                inputs.flag_options.cstruct,
+            )
+        else:
+            lib.Broadcast_struct_global_noastro(
+                inputs.user_params.cstruct,
+                inputs.cosmo_params.cstruct,
+            )
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def init_backend_ps(func: Callable) -> Callable:
+    """Decorator to initialise the backend PS before calling the function."""
+
+    @broadcast_params
+    def wrapper(*args, **kwargs):
+        lib.init_ps()
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def init_sigma_table(func: Callable) -> Callable:
+    """Decorator to initialise the the sigma interpolation table before calling the function."""
+
+    @init_backend_ps
+    def wrapper(*args, **kwargs):
+        sigma_min_mass = kwargs.get("M_min", 1e5)
+        sigma_max_mass = kwargs.get("M_max", 1e16)
+        if (
+            kwargs.get("inputs").user_params.USE_INTERPOLATION_TABLES
+            != "no-interpolation"
+        ):
+            lib.initialiseSigmaMInterpTable(sigma_min_mass, sigma_max_mass)
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+def init_gl(func: Callable) -> Callable:
+    """Decorator to initialise the Gauss-Legendre if required before calling the function."""
```

Comment:
I'm not sure what "initialise the Gauss-Legendre" means -- is this setting up GSL integration methods?